WordPress/gutenberg
# Scripts: Optimize default Webpack configuration

This pull request seeks to try a few optimizations for the default Webpack configuration. In my testing using `time ./node_modules/.bin/wp-scripts build`, the approximate result in my environment is a net build time decrease of ~38% (18.5s to 11.5s) uncached to ~50% (18.5s to 9.3s) cached.

The optimizations are three:

- Use [`thread-loader`](https://github.com/webpack-contrib/thread-loader) to complement Babel compilation. On its own, with a default configuration, this seemed to reduce build time by approximately 2-3 seconds. The result may vary by machine depending on the number of cores available (default threading is `numCPU - 1` workers).
- Opt-in to [`babel-loader`'s `cacheDirectory` option](https://webpack.js.org/loaders/babel-loader/#options). This too appeared to reduce build time by approximately 2-3 seconds further. By default, this uses a cache directory local to the project's `node_modules`. Thus, it will not have an immediate benefit on a fresh clone, or if `node_modules` is removed for any reason. I've included an option to direct this to a more persistent cache directory, which is partly self-serving for one of my [own projects](https://github.com/aduth/gutenberg.run) where I'd like to use a more persistent cache.
- Exclude `source-map-loader` from production builds. This had a slightly more significant impact to reduce build time by a further 3-4 seconds. My reading of the [`source-map-loader` documentation](https://webpack.js.org/loaders/source-map-loader/) is that it is intended to complement the `devtool` option, which we already [configure only for non-production environments](https://github.com/WordPress/gutenberg/blob/a9f6a1e0eee428831a9482e0d3c36ee9e4004056/packages/scripts/config/webpack.config.js#L120-L124).
   - >All source map data is passed to webpack for processing as per a chosen source map style specified by the `devtool` option in webpack.config.js.

A few others I considered, but did not include:

- Updating Webpack to the latest version. Webpack 5.x promises performance improvements, but is currently in alpha, and thus I did not consider stable enough to explore here, or at least not as part of this pull request. The latest minor/patch version combination actually appeared to _increase_ build time.
- A few other tips from [Webpack's Build Performance guide](https://webpack.js.org/guides/build-performance) which did not appear to have much impact positive or negative (`output.pathinfo`).

**References:**

- https://slack.engineering/keep-webpack-fast-a-field-guide-for-better-build-performance-f56a5995e8f1
- https://webpack.js.org/guides/build-performance/
- https://blog.box.com/blog/how-we-improved-webpack-build-performance-95

**Testing instructions:**

You may consider to compare build times between master and this branch:

```
time ./node_modules/.bin/wp-scripts build
```

_(Note: Babel loader's `cacheDirectory` won't have an impact the first time it's run. You can look to see whether a cache exists as `node_modules/.cache/babel-loader`)_

Otherwise, verify there are no regressions in the expected output from build scripts (runs in browser, no sourcemap/devtool regression).
